qiskit/operations.py

```diff
@@ -0,0 +1,321 @@
+from qiskit.quantum_info import Operator
+
+from qiskit.circuit import QuantumCircuit
+
+from qiskit.result.counts import Counts
+from qiskit.result import QuasiDistribution
+
+from qiskit._accelerate import pauli_expval
+
+# type of collection of qubit indices to measure
+QargsT = Union[list, range]
+
+
+###
+### Make a uniform interface to a couple of properties of Statevector and DensityMatrix
+###
+
+
+@dispatch
+def _to_rust_array(state: Statevector):
+    return state.data
+
+
+@dispatch
+def _to_rust_array(state: DensityMatrix):
+    return np.ravel(state.data, order="F")
+
+
+@dispatch
+def sum_of_probs(state: Statevector):
+    return np.linalg.norm(state.data)
+
+
+@dispatch
+def sum_of_probs(state: DensityMatrix):
+    return state.trace()
+
+
+###
+### Make a uniform interface `expval_pauli` to the four rust routines.
+###
+
+
+@dispatch
+def expval_pauli(state: Statevector, z_mask):
+    return pauli_expval.expval_pauli_no_x(_to_rust_array(state), state.num_qubits, z_mask)
+
+
+@dispatch
+def expval_pauli(state: Statevector, z_mask, x_mask, y_phase, x_max):
+    return pauli_expval.expval_pauli_with_x(
+        _to_rust_array(state), state.num_qubits, z_mask, x_mask, y_phase, x_max
+    )
+
+
+@dispatch
+def expval_pauli(state: DensityMatrix, z_mask):
+    data = _to_rust_array(state)
+    return pauli_expval.density_expval_pauli_no_x(data, state.num_qubits, z_mask)
+
+
+@dispatch
+def expval_pauli(state: DensityMatrix, z_mask, x_mask, y_phase, x_max):
+    data = _to_rust_array(state)
+    return pauli_expval.density_expval_pauli_with_x(
+        data, state.num_qubits, z_mask, x_mask, y_phase, x_max
+    )
+
+
+# This is currently used twice.
+def pauli_phase(pauli: Pauli):
+    return (-1j) ** pauli.phase if pauli.phase else 1
+
+
+## Function `expectation_value`.
+## expectation_value(oper, state)
+## expectation_value(oper, state, qargs) # to restrict to a subset of qubits
+## expectation_value(state) # if the operator is implied
+## expectation_value(state, qargs) # operator implied, restricted subset of qubits
+
+## Compare this to the implementations in quantum_info classes.
+## Here, we have a single method for Statevector, DensityMatrix rather than two.
+## Furthermore, the internal structure of `state` is not accessed in this method.
+@dispatch(precedence=1)
+def expectation_value(pauli: Pauli, state: Union[Statevector, DensityMatrix], qargs: QargsT):
+
+    qubits = np.array(qargs)
+    x_mask = np.dot(1 << qubits, pauli.x)
+    z_mask = np.dot(1 << qubits, pauli.z)
+    _pauli_phase = pauli_phase(pauli)
+
+    if x_mask + z_mask == 0:
+        return _pauli_phase * sum_of_probs(state)
+
+    if x_mask == 0:
+        return _pauli_phase * expval_pauli(state, z_mask)
+
+    x_max = qubits[pauli.x][-1]
+    y_phase = (-1j) ** pauli._count_y()
+    return _pauli_phase * expval_pauli(state, z_mask, x_mask, y_phase, x_max)
+
+
+@dispatch(precedence=1)
+def expectation_value(oper: Pauli, state: StabilizerState, qargs: QargsT):
```

Comment:
That makes sense. In particular in the case of `StabilizerState` (as far as this PR is developed). But `Statevector/DensityMatrix` share a single method here. Which class should this go in? Or maybe I'm misunderstanding something.

How to integrate MD with existing language features and styles is not clear to me. There is an argument for doing it the other way around. Have the class methods call the generic functions. Or maybe its better to avoid being able call class methods and generic functions to do the same thing.